common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -22,64 +22,311 @@
+                @Override
+                public void invoke(Result<UsbSmartCardConnection, IOException> value) {
+                    try {
+                        final SmartCardConnection c = value.getValue();
+                        PivSession piv = new PivSession(c);
+                        //Check if too many PIN attempts have been made.
+                        //If so, go directly to Error dialog.
+                        if (piv.getPinAttempts() == 0) {
+                            SmartcardMaxFailedAttemptsDialog maxFailedAttemptsDialog = new SmartcardMaxFailedAttemptsDialog(mActivity);
+                            maxFailedAttemptsDialog.show();
+                            request.cancel();
+                            Logger.infoPII(methodTag,  "User has reached the maximum failed attempts allowed.");
+                            return;
+                        }
+                        //Create ArrayList that contains cert details only pertinent to the cert picker.
+                        final ArrayList<YubiKitCertDetails> certList = new ArrayList<>();
+                        //We need to check every slot. If a slot is empty, an ApduException is called, in which case, we'll ignore it.
+                        //AUTHENTICATION (9A)
+                        getAndPutCertDetailsInList(AUTHENTICATION, piv, certList);
+                        //SIGNATURE (9C)
+                        getAndPutCertDetailsInList(SIGNATURE, piv, certList);
+                        //KEY_MANAGEMENT (9D)
+                        getAndPutCertDetailsInList(KEY_MANAGEMENT, piv, certList);
+                        //CARD_AUTH (9E)
+                        getAndPutCertDetailsInList(CARD_AUTH, piv, certList);
+                        //If no certs were found, cancel flow.
+                        if (certList.isEmpty()) {
+                            request.cancel();
+                            Logger.infoPII(methodTag,  "No PIV certificates found on YubiKey device.");
+                            return;
+                        }
+                        //Show Smartcard cert picker, which also handles the rest of the smartcard CBA flow.
+                        final SmartcardCertPickerDialog certPickerDialog = new SmartcardCertPickerDialog(certList, mActivity);
+                        certPickerDialog.setPositiveButtonListener(getSmartcardCertPickerDialogPositiveButtonListener(initialDevice, request));
+                        certPickerDialog.setNegativeButtonListener(new SmartcardCertPickerDialog.NegativeButtonListener() {
+                            @Override
+                            public void onClick() {
+                                request.cancel();
+                            }
+                        });
+
+                        certPickerDialog.show();
+
+                    } catch(IOException e) {
+                        Logger.errorPII(methodTag, "IOException", e);
```

Comment:
What would be the user experience if we run into any of these exceptions?

Reply:
The ClientCertRequest object should have 'cancel' called on it in these cases... which will cancel the flow and show the user an error message. I'll add these lines in.

Reply:
Should we display an error message to the user?

Reply:
Would it be helpful for the user to see the error message? I was thinking that in these types of exceptions, it would be more helpful for the developers.. unless it would also be more helpful for us to do more than just error log.

Reply:
@melissaahn probably. I would be confused if I plug yubikey in and i'm seeing the same error as if I don't plug it in.

Reply:
I'll look into creating an error dialog for these cases for next PR

Reply:
let's add TODO comment in these blocks as a reminder

```Java
// TODO: XXXXXXXX
```